src/main/java/duckling/Server.java

```diff
@@ -1,11 +1,101 @@
 package duckling;
 
+import duckling.requests.Request;
+import duckling.requests.RequestBuilder;
+import duckling.responders.Responders;
+
+import java.io.*;
+import java.net.*;
+import java.util.concurrent.*;
+
 public class Server {
+    public static final String CLRF = "\r\n";
+
     public int port;
+    public String root;
+
+    private boolean shuttingDown = false;
```

Comment:
I think this variable should be `volatile`: that way, Java would guarantee your writes to that variable will be visible across threads as expected, and that your loop doesn't turn into an infinite loop by the JIT. This description is pretty dense but covers parts of the issue: http://tutorials.jenkov.com/java-concurrency/volatile.html And this SO question is basically exactly the issue I'm talking about: http://stackoverflow.com/questions/37348390/java-jit-compiler-optimizations-is-jit-consistent-in-respect-to-volatile-varia

[Java Concurrency in Practice](http://jcip.net/) is a terrific book on concurrency and I think it's helpful across languages to see where issues might arise.